File: frontendios/src/services/userMetricsService.ts
/**
 * Service User Metrics pour EdgeCoach iOS
 * Gestion des métriques personnelles utilisateur (poids, FC, objectifs, etc.)
 */

import apiService from './api';

// Types
export interface UserMetricsData {
  height: string;
  weight: string;
  weeklyVolume: string;
  primaryGoal: string;
  availableDays: string;
  preferredTrainingTime: string;
  equipment: string;
  medicalConditions: string;
  ftp: string;
  lthr: string;
  maxHr: string;
  restHr: string;
  vo2Max: string;
}

export interface ApiUserMetrics {
  height?: number;
  weight?: number;
  weekly_volume?: number;
  primary_goal?: string;
  available_days?: number;
  preferred_training_time?: string;
  equipment?: string;
  medical_conditions?: string;
  ftp?: number;
  lthr?: number;
  max_hr?: number;
  rest_hr?: number;
  vo2_max?: number;
}

export interface UserMetricsResult {
  success: boolean;
  metrics?: UserMetricsData;
  message?: string;
  error?: string;
}

export interface MetricsSummary {
  totalWorkouts: number;
  totalDistance: number;
  totalDuration: number;
  averageHeartRate: number;
  weeklyProgress: number;
}

export interface MetricsSummaryResult {
  success: boolean;
  summary?: MetricsSummary;
  error?: string;
}

class UserMetricsService {
  /**
   * Récupérer les métriques d'un utilisateur
   */
  async getMetrics(userId: string): Promise<UserMetricsResult> {
    try {
      const response = await apiService.get<ApiUserMetrics>(`/users/${userId}/metrics`);
      const metrics = this.convertApiToFrontend(response);

      return {
        success: true,
        metrics,
      };
    } catch (error: any) {
      return {
        success: false,
        error: error.message,
      };
    }
  }

  /**
   * Créer ou mettre à jour les métriques utilisateur
   */
  async updateMetrics(userId: string, metricsData: Partial<UserMetricsData>): Promise<UserMetricsResult> {
    try {
      // Convertir les données du format frontend vers le format API
      const apiData: ApiUserMetrics = {};

      if (metricsData.height) apiData.height = parseFloat(metricsData.height);
      if (metricsData.weight) apiData.weight = parseFloat(metricsData.weight);
      if (metricsData.weeklyVolume) apiData.weekly_volume = parseFloat(metricsData.weeklyVolume);
      if (metricsData.primaryGoal) apiData.primary_goal = metricsData.primaryGoal;
      if (metricsData.availableDays) apiData.available_days = parseInt(metricsData.availableDays, 10);
      if (metricsData.preferredTrainingTime) apiData.preferred_training_time = metricsData.preferredTrainingTime;
      if (metricsData.equipment) apiData.equipment = metricsData.equipment;
      if (metricsData.medicalConditions) apiData.medical_conditions = metricsData.medicalConditions;
      if (metricsData.ftp) apiData.ftp = parseFloat(metricsData.ftp);
      if (metricsData.lthr) apiData.lthr = parseInt(metricsData.lthr, 10);
      if (metricsData.maxHr) apiData.max_hr = parseInt(metricsData.maxHr, 10);
      if (metricsData.restHr) apiData.rest_hr = parseInt(metricsData.restHr, 10);
      if (metricsData.vo2Max) apiData.vo2_max = parseFloat(metricsData.vo2Max);

      const response = await apiService.put<{ metrics: ApiUserMetrics; message: string }>(
        `/users/${userId}/metrics`,
        apiData
      );

      return {
        success: true,
        metrics: this.convertApiToFrontend(response.metrics),
        message: response.message,
      };
    } catch (error: any) {
      return {
        success: false,
        error: error.message,
      };
    }
  }

  /**
   * Récupérer un résumé des métriques utilisateur
   */
  async getMetricsSummary(userId: string): Promise<MetricsSummaryResult> {
    try {
      const response = await apiService.get<MetricsSummary>(`/users/${userId}/metrics/summary`);

      return {
        success: true,
        summary: response,
      };
    } catch (error: any) {
      return {
        success: false,
        error: error.message,
      };
    }
  }

  /**
   * Convertir les données API vers le format frontend
   */
  convertApiToFrontend(apiMetrics: ApiUserMetrics | undefined): UserMetricsData {
    if (!apiMetrics) {
      return {
        height: '',
        weight: '',
        weeklyVolume: '',
        primaryGoal: '',
        availableDays: '',
        preferredTrainingTime: '',
        equipment: '',
        medicalConditions: '',
        ftp: '',
        lthr: '',
        maxHr: '',
        restHr: '',
        vo2Max: '',
      };
    }

    return {
      height: apiMetrics.height ? apiMetrics.height.toString() : '',
      weight: apiMetrics.weight ? apiMetrics.weight.toString() : '',
      weeklyVolume: apiMetrics.weekly_volume ? apiMetrics.weekly_volume.toString() : '',
      primaryGoal: apiMetrics.primary_goal || '',
      availableDays: apiMetrics.available_days ? apiMetrics.available_days.toString() : '',
      preferredTrainingTime: apiMetrics.preferred_training_time || '',
      equipment: apiMetrics.equipment || '',
      medicalConditions: apiMetrics.medical_conditions || '',
      ftp: apiMetrics.ftp ? apiMetrics.ftp.toString() : '',
      lthr: apiMetrics.lthr ? apiMetrics.lthr.toString() : '',
      maxHr: apiMetrics.max_hr ? apiMetrics.max_hr.toString() : '',
      restHr: apiMetrics.rest_hr ? apiMetrics.rest_hr.toString() : '',
      vo2Max: apiMetrics.vo2_max ? apiMetrics.vo2_max.toString() : '',
    };
  }
}

export const userMetricsService = new UserMetricsService();
export default userMetricsService;

File: EdgeCoachSwiftUI/EdgeCoach/Services/UserMetricsService.swift
/**
 * Service User Metrics pour EdgeCoach iOS
 * Gestion des métriques personnelles utilisateur (poids, FC, objectifs, etc.)
 * Aligné avec frontendios/src/services/userMetricsService.ts
 */

import Foundation

// MARK: - User Metrics Models

struct UserMetricsData: Codable {
    var height: String
    var weight: String
    var weeklyVolume: String
    var primaryGoal: String
    var availableDays: String
    var preferredTrainingTime: String
    var equipment: String
    var medicalConditions: String
    var ftp: String
    var lthr: String
    var maxHr: String
    var restHr: String
    var vo2Max: String

    static func empty() -> UserMetricsData {
        return UserMetricsData(
            height: "",
            weight: "",
            weeklyVolume: "",
            primaryGoal: "",
            availableDays: "",
            preferredTrainingTime: "",
            equipment: "",
            medicalConditions: "",
            ftp: "",
            lthr: "",
            maxHr: "",
            restHr: "",
            vo2Max: ""
        )
    }
}

struct APIUserMetrics: Codable {
    let height: Double?
    let weight: Double?
    let weeklyVolume: Double?
    let primaryGoal: String?
    let availableDays: Int?
    let preferredTrainingTime: String?
    let equipment: String?
    let medicalConditions: String?
    let ftp: Double?
    let lthr: Int?
    let maxHr: Int?
    let restHr: Int?
    let vo2Max: Double?

    enum CodingKeys: String, CodingKey {
        case height
        case weight
        case weeklyVolume = "weekly_volume"
        case primaryGoal = "primary_goal"
        case availableDays = "available_days"
        case preferredTrainingTime = "preferred_training_time"
        case equipment
        case medicalConditions = "medical_conditions"
        case ftp
        case lthr
        case maxHr = "max_hr"
        case restHr = "rest_hr"
        case vo2Max = "vo2_max"
    }
}

struct MetricsSummary: Codable {
    let totalWorkouts: Int
    let totalDistance: Double
    let totalDuration: Int
    let averageHeartRate: Int
    let weeklyProgress: Double

    enum CodingKeys: String, CodingKey {
        case totalWorkouts = "total_workouts"
        case totalDistance = "total_distance"
        case totalDuration = "total_duration"
        case averageHeartRate = "average_heart_rate"
        case weeklyProgress = "weekly_progress"
    }
}

// MARK: - User Metrics Service

@MainActor
class UserMetricsService {
    static let shared = UserMetricsService()
    private let api = APIService.shared

    private init() {}

    // MARK: - Get Metrics

    func getMetrics(userId: String) async throws -> UserMetricsData {
        do {
            let response: APIUserMetrics = try await api.get("/users/\(userId)/metrics")
            return convertApiToFrontend(response)
        } catch {
            #if DEBUG
            print("Failed to get user metrics: \(error)")
            #endif
            return UserMetricsData.empty()
        }
    }

    // MARK: - Update Metrics

    func updateMetrics(userId: String, metricsData: UserMetricsData) async throws -> UserMetricsData {
        struct UpdateResponse: Decodable {
            let metrics: APIUserMetrics?
            let message: String?
        }

        var apiData: [String: Any] = [:]

        if !metricsData.height.isEmpty, let value = Double(metricsData.height) {
            apiData["height"] = value
        }
        if !metricsData.weight.isEmpty, let value = Double(metricsData.weight) {
            apiData["weight"] = value
        }
        if !metricsData.weeklyVolume.isEmpty, let value = Double(metricsData.weeklyVolume) {
            apiData["weekly_volume"] = value
        }
        if !metricsData.primaryGoal.isEmpty {
            apiData["primary_goal"] = metricsData.primaryGoal
        }
        if !metricsData.availableDays.isEmpty, let value = Int(metricsData.availableDays) {
            apiData["available_days"] = value
        }
        if !metricsData.preferredTrainingTime.isEmpty {
            apiData["preferred_training_time"] = metricsData.preferredTrainingTime
        }
        if !metricsData.equipment.isEmpty {
            apiData["equipment"] = metricsData.equipment
        }
        if !metricsData.medicalConditions.isEmpty {
            apiData["medical_conditions"] = metricsData.medicalConditions
        }
        if !metricsData.ftp.isEmpty, let value = Double(metricsData.ftp) {
            apiData["ftp"] = value
        }
        if !metricsData.lthr.isEmpty, let value = Int(metricsData.lthr) {
            apiData["lthr"] = value
        }
        if !metricsData.maxHr.isEmpty, let value = Int(metricsData.maxHr) {
            apiData["max_hr"] = value
        }
        if !metricsData.restHr.isEmpty, let value = Int(metricsData.restHr) {
            apiData["rest_hr"] = value
        }
        if !metricsData.vo2Max.isEmpty, let value = Double(metricsData.vo2Max) {
            apiData["vo2_max"] = value
        }

        // Create encodable struct dynamically
        struct MetricsUpdate: Encodable {
            let height: Double?
            let weight: Double?
            let weeklyVolume: Double?
            let primaryGoal: String?
            let availableDays: Int?
            let preferredTrainingTime: String?
            let equipment: String?
            let medicalConditions: String?
            let ftp: Double?
            let lthr: Int?
            let maxHr: Int?
            let restHr: Int?
            let vo2Max: Double?

            enum CodingKeys: String, CodingKey {
                case height, weight
                case weeklyVolume = "weekly_volume"
                case primaryGoal = "primary_goal"
                case availableDays = "available_days"
                case preferredTrainingTime = "preferred_training_time"
                case equipment
                case medicalConditions = "medical_conditions"
                case ftp, lthr
                case maxHr = "max_hr"
                case restHr = "rest_hr"
                case vo2Max = "vo2_max"
            }
        }

        let update = MetricsUpdate(
            height: Double(metricsData.height),
            weight: Double(metricsData.weight),
            weeklyVolume: Double(metricsData.weeklyVolume),
            primaryGoal: metricsData.primaryGoal.isEmpty ? nil : metricsData.primaryGoal,
            availableDays: Int(metricsData.availableDays),
            preferredTrainingTime: metricsData.preferredTrainingTime.isEmpty ? nil : metricsData.preferredTrainingTime,
            equipment: metricsData.equipment.isEmpty ? nil : metricsData.equipment,
            medicalConditions: metricsData.medicalConditions.isEmpty ? nil : metricsData.medicalConditions,
            ftp: Double(metricsData.ftp),
            lthr: Int(metricsData.lthr),
            maxHr: Int(metricsData.maxHr),
            restHr: Int(metricsData.restHr),
            vo2Max: Double(metricsData.vo2Max)
        )

        let response: UpdateResponse = try await api.put("/users/\(userId)/metrics", body: update)

        if let metrics = response.metrics {
            return convertApiToFrontend(metrics)
        }

        return metricsData
    }

    // MARK: - Get Summary

    func getMetricsSummary(userId: String) async throws -> MetricsSummary {
        return try await api.get("/users/\(userId)/metrics/summary")
    }

    // MARK: - Aggregated Metrics

    func getAggregatedMetrics(userId: String) async throws -> TrainingLoad? {
        struct Response: Codable {
            let meta: Meta?
            
            struct Meta: Codable {
                let trainingLoad: TrainingLoad?
                
                enum CodingKeys: String, CodingKey {
                    case trainingLoad = "training_load"
                }
            }
        }

        let response: Response = try await api.get(
            "/users/\(userId)/aggregate-metrics",
            queryParams: ["use_cache": "true"]
        )
        return response.meta?.trainingLoad
    }

    // MARK: - Convert API to Frontend

    private func convertApiToFrontend(_ apiMetrics: APIUserMetrics?) -> UserMetricsData {
        guard let api = apiMetrics else {
            return UserMetricsData.empty()
        }

        return UserMetricsData(
            height: api.height.map { String($0) } ?? "",
            weight: api.weight.map { String($0) } ?? "",
            weeklyVolume: api.weeklyVolume.map { String($0) } ?? "",
            primaryGoal: api.primaryGoal ?? "",
            availableDays: api.availableDays.map { String($0) } ?? "",
            preferredTrainingTime: api.preferredTrainingTime ?? "",
            equipment: api.equipment ?? "",
            medicalConditions: api.medicalConditions ?? "",
            ftp: api.ftp.map { String($0) } ?? "",
            lthr: api.lthr.map { String($0) } ?? "",
            maxHr: api.maxHr.map { String($0) } ?? "",
            restHr: api.restHr.map { String($0) } ?? "",
            vo2Max: api.vo2Max.map { String($0) } ?? ""
        )
    }
}
